### PR TITLE
More constexpr interpolator, separate dual/primal weights/startIndices

### PR DESCRIPTION
**Pull request by PhilipDeegan**

### CODE REVIEW

File: src/core/numerics/interpolator/interpolator.hpp [hunk 1]
```diff
@@ -12,677 +12,598 @@
...
+        auto const& [Bx, By, Bz] = Em.B();
 
+        // for each particle, first calculate the startIndex and weights for dual and
+        // primal quantities. then, knowing the centering (primal or dual) of each
+        // electromagnetic component, we use Interpol to actually perform the
+        // interpolation. the trick here is that the StartIndex and weights have only been
+        // calculated twice, and not for each E,B component.
 
+        PHARE_LOG_START("MeshToParticle::operator()");
+        std::size_t eb_idx = 0;
```
Comment (nicolasaunai) [inline src/core/numerics/interpolator/interpolator.hpp:488]: unused ?

File: src/core/numerics/interpolator/interpolator.hpp [hunk 2]
```diff
@@ -12,677 +12,598 @@
...
+        auto const& [xFlux, yFlux, zFlux]      = flux();
+        auto const& [xStartIndex, yStartIndex] = startIndex;
+        auto const& [xWeights, yWeights]       = weights;
+        auto const& order_size                 = xWeights.size();
 
-            auto const partRho   = particle.weight;
-            auto const xPartFlux = particle.v[0] * particle.weight;
-            auto const yPartFlux = particle.v[1] * particle.weight;
-            auto const zPartFlux = particle.v[2] * particle.weight;
+        auto const partRho   = particle.weight * coef;
```
[the remark this answers is not in the document]
Comment (PhilipDeegan) [reply src/core/numerics/interpolator/interpolator.hpp:345]: forgot

File: src/core/numerics/interpolator/interpolator.hpp [hunk 3]
```diff
@@ -12,677 +12,598 @@
...
+//! Interpol performs the interpolation of a field using precomputed weights at
+//! indices starting at startIndex. The class is templated by the Dimensionality
+template<std::size_t dim>
+class MeshToParticle
+{
+};
 
 
-    /** \brief specialization of Interpol for 1D interpolation
+template<std::size_t dimdex, typename GridLayout, typename Quantity, Quantity quantity,
```
[the remark this answers is not in the document]
Comment (PhilipDeegan) [reply src/core/numerics/interpolator/interpolator.hpp:136]: yes

**Comment by PhilipDeegan** [comment]
> looks ok to me. Few questions. Also, curiosity : did you find any noticeable speedup with these changes?

I haven't investigated too deeply for AoS, but we can

**Review by nicolasaunai — APPROVED** [review]
thx

**Pull request merged**